src/service/xhr-impl.js

```diff
@@ -514,61 +198,89 @@ export class Xhr {
    */
   sendSignal(input, opt_init) {
     return this.fetchAmpCors_(input, opt_init)
-        .then(response => assertSuccess(response));
+        .then(res => {
+          const response = /**@type {!Response} */ (res);
+          return assertSuccess(response);
+        });
   }
 
   /**
-   * Add "__amp_source_origin" query parameter to the URL. Ideally, we'd be
-   * able to set a header (e.g. AMP-Source-Origin), but this will force
-   * preflight request on all CORS request.
-   * @param {!Window} win
-   * @param {string} url
-   * @return {string}
+   * @override
    */
-  getCorsUrl(win, url) {
-    return getCorsUrl(win, url);
+  fromStructuredCloneable_(response, responseType) {
+    user().assert(isObject(response), 'Object expected: %s', response);
+
+    const isDocumentType = responseType == 'document';
+    if (typeof this.win.Response === 'function' && !isDocumentType) {
+      // Use native `Response` type if available for performance. If response
+      // type is `document`, we must fall back to `FetchResponse` polyfill
+      // because callers would then rely on the `responseXML` property being
+      // present, which is not supported by the Response type.
+      return new this.win.Response(response['body'], response['init']);
+    }
+
+    // TODO(prateekbh): shift below loginc into document-fetch
+    // after fetch-polyfill.
+    const lowercasedHeaders = map();
+    const data = {
+      status: 200,
+      statusText: 'OK',
+      responseText: (response['body'] ? String(response['body']) : ''),
+      /**
+       * @param {string} name
+       * @return {string}
+       */
+      getResponseHeader(name) {
+        return lowercasedHeaders[String(name).toLowerCase()] || null;
+      },
+    };
+
+    if (response['init']) {
+      const init = response['init'];
+      if (isArray(init.headers)) {
+        init.headers.forEach(entry => {
+          const headerName = entry[0];
+          const headerValue = entry[1];
+          lowercasedHeaders[String(headerName).toLowerCase()] =
+              String(headerValue);
+        });
+      }
+      if (init.status) {
+        data.status = parseInt(init.status, 10);
+      }
+      if (init.statusText) {
+        data.statusText = String(init.statusText);
+      }
+    }
+
+    if (isDocumentType) {
```

Comment:
This shouldn't be possible anymore, all callers should be going through `DocumentFetcher`.

Reply:
Not in this PR(otherwise this will become a gigantic PR of 20-25 files with type changes polyfills etc). This PR just creates the `document-fetcher` nobody uses it as of this PR.

The 4 PRs will be: https://github.com/ampproject/amphtml/issues/16707